x-pack/plugins/security/server/lib/authorization/check_privileges.js

```diff
@@ -4,103 +4,82 @@
+    return CHECK_PRIVILEGES_RESULT.UNAUTHORIZED;
+  };
+
+  return function checkPrivilegesWithRequest(request) {
+
+    return async function checkPrivileges(privileges) {
+      const allApplicationPrivileges = uniq([actions.version, actions.login, ...privileges]);
+      const hasPrivilegesResponse = await callWithRequest(request, 'shield.hasPrivileges', {
         body: {
+          applications: [{
+            application,
+            resources: [ALL_RESOURCE],
+            privileges: allApplicationPrivileges
+          }],
           index: [{
             names: [kibanaIndex],
             privileges: ['create', 'delete', 'read', 'view_index_metadata']
-          }]
+          }],
         }
       });
 
-      return Object.values(privilegeCheck.index[kibanaIndex]).includes(true);
-    };
-
-    return async function checkPrivileges(privileges) {
-      const allPrivileges = [versionAction, loginAction, ...privileges];
-      const applicationPrivilegesCheck = await checkApplicationPrivileges(allPrivileges);
-
-      const username = applicationPrivilegesCheck.username;
-
-      // We don't want to expose the version privilege to consumers, as it's an implementation detail only to detect version mismatch
-      const missing = Object.keys(applicationPrivilegesCheck.privileges)
-        .filter(p => !applicationPrivilegesCheck.privileges[p])
-        .filter(p => p !== versionAction);
-
-      if (applicationPrivilegesCheck.hasAllRequested) {
-        return {
-          result: CHECK_PRIVILEGES_RESULT.AUTHORIZED,
-          username,
-          missing,
-        };
-      }
+      const applicationPrivilegesResponse = hasPrivilegesResponse.application[application][ALL_RESOURCE];
+      const indexPrivilegesResponse =  hasPrivilegesResponse.index[kibanaIndex];
 
-      if (!applicationPrivilegesCheck.privileges[loginAction] && await hasPrivilegesOnKibanaIndex()) {
-        const msg = 'Relying on index privileges is deprecated and will be removed in Kibana 7.0';
-        server.log(['warning', 'deprecated', 'security'], msg);
-
-        return {
-          result: CHECK_PRIVILEGES_RESULT.LEGACY,
-          username,
-          missing,
-        };
+      if (hasIncompatibileVersion(applicationPrivilegesResponse)) {
+        throw new Error('Multiple versions of Kibana are running against the same Elasticsearch cluster, unable to authorize user.');
       }
 
       return {
-        result: CHECK_PRIVILEGES_RESULT.UNAUTHORIZED,
-        username,
-        missing,
+        result: determineResult(applicationPrivilegesResponse, indexPrivilegesResponse),
+        username: hasPrivilegesResponse.username,
+
+        // we only return missing privileges that they're specifically checking for
+        missing: Object.keys(applicationPrivilegesResponse)
+          .filter(privilege => privileges.includes(privilege))
+          .filter(privilege => !applicationPrivilegesResponse[privilege])
```

Comment:
nit: These `filter` calls could probably be combined without really sacrificing readability